build_tools/benchmarks/run_benchmarks_on_android.py

```diff
@@ -209,18 +282,28 @@ def run_benchmark_case(
         benchmark_results_filename: Optional[pathlib.Path],
         capture_filename: Optional[pathlib.Path],
     ) -> None:
-        benchmark_case_dir = benchmark_case.benchmark_case_dir
-        android_case_dir = pathlib.PurePosixPath(
-            benchmark_case_dir.relative_to(self.config.root_benchmark_dir)
+        module_rel_dir = iree_artifacts.get_module_dir_path(
+            benchmark_case.run_config.module_generation_config
         )
-
-        self.__check_and_push_file(
-            benchmark_case_dir / iree_artifacts.MODULE_FILENAME, android_case_dir
+        android_case_dir = ANDROID_TMPDIR / module_rel_dir
+
+        module_dir = benchmark_case.module_dir
+        if isinstance(module_dir, pathlib.Path):
+            module_path = module_dir / iree_artifacts.MODULE_FILENAME
+        else:
+            module_path = urllib.parse.urljoin(
+                module_dir, iree_artifacts.MODULE_FILENAME
+            )
```

Comment:
There are a few places with this switch on the loose Python types. You could introduce a new class like `DirectoryLocation` that centralizes that logic. This code feels a bit error prone (could forget to check if "module_dir" is an instance of `pathlib.Path` in one file).

WDYT?

Reply:
The `pyright/pytype` actually helps a lot to check this kind of error IMO. They will complain when you forgot to handle the union types properly. But we are actually able to fold some nontrivial logic into a new class.

Added the `ResourceLocation` for this.